torch/utils/data/distributed.py

```diff
@@ -65,3 +65,117 @@ def __len__(self):
 
     def set_epoch(self, epoch):
         self.epoch = epoch
+
+class ChunkDataReader(object):
+    r"""Reads a chunk of data given a chunk index.
+
+    A chunk could be full file, section of a large file, folders, URLs, or any other
+    abstraction that allows data to be segmented roughtly the same size.
+
+    As an example, chunking could be used in a scenario where MNIST dataset for training 
+    would be converted in CSV text files and equally splitted in several files,
+    e.g. 60 files with 1000 lines each.  In this case, each worker would read
+    a text file with all 1000 images.
+
+    Another example would be chunks as individual folders, containing multiple audio files.
+    In this scenario, each worker would have access to different folders, 
+    where binary files would be read from the filesystem.
+
+    A chunk doesn necessarily need to rely in multiple files to split data laoding across workers.
+    It is possible to extract chunks from a single large dataset file, where each worker would
+    seek to specific positions to load data.
+
+    The reading logic must be implemented inside `__call__` method
+    """
+
+    def __init__(self):
+        pass
+
+    def __call__(self, idx):
+        r"""Returns `list(samples)`
+
+        If `samples` contain tensors, numpy arrays, numbers,
+        dicts or lists, the default collate function can be implicitly used.
+        Otherwise, a custom `collate_fn` must be provided to `DataLoader`
+
+        When no more data is available, `StopIteration` must be raised.
+        """
+        raise NotImplementedError
+
+class DistributedChunkSampler(Sampler):
+    r"""This sampler introduces distributed sampling without padding and dataset dependency.
+
```

Comment:
This sampler very similar to the DistributedSampler, however avoid padding and the dependency on the size of the dataset. With two levels of sampling, the DisributedChunkSampler is used by the dataset and hence run by a dataloader worker. To distinguish workers from the same node and distribute data among all participating workers, each sampler needs to be configured with global_worker_rank using the worker_initialization function.